### PR TITLE
Add additional exclusions for dependency collection

### PR DESCRIPTION
**Pull request by andrewlock**
## Summary of changes

Excludes 32-long hex strings with a 0.0.0.0 version number

## Reason for change

We recently saw an explosion of these (we're not sure exactly what's generating them) 

## Implementation details

Add additional exclusion pattern for 0 version dependencies

## Test coverage

Added some test cases

## Other details

We could make this more general (i.e. don't worry about the 0.0.0.0 version bit), my concern is excluding "valid" assemblies accidentally

### CODE REVIEW

**Comment by andrewlock** [comment]
## Execution-Time Benchmarks Report :stopwatch:

Execution-time results for samples comparing the following branches/commits:
- [This PR (6282)](https://github.com/DataDog/dd-trace-dotnet/tree/42843f51a5c9f9fcee7ec09554ca49e4871ff160)
- [master](https://github.com/DataDog/dd-trace-dotnet/tree/6aa8c666b2a48b6c5a85044ce60a31a23c80970e)

Execution-time benchmarks measure the whole time it takes to execute a program. And are intended to measure the one-off costs. Cases where the execution time results for the PR are worse than latest master results are shown in **red**. The following thresholds were used for comparing the execution times:
* Welch test with statistical test for significance of **5%**
* Only results indicating a difference greater than **5%** and **5 ms** are considered.


Note that these results are based on a _single_ point-in-time result for each branch. For full results, see the [dashboard](https://ddstaging.datadoghq.com/dashboard/4qn-6fi-54p/apm-net-execution-time-benchmarks).

Graphs show the p99 interval based on the mean and StdDev of the test run, as well as the mean value of the run (shown as a diamond below the graph).
```mermaid
gantt
    title Execution time (ms) FakeDbCommand (.NET Framework 4.6.2) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (6282) - mean (73ms)  : 61, 85
     .   : milestone, 73,
    master - mean (72ms)  : 63, 82
     .   : milestone, 72,

    section CallTarget+Inlining+NGEN
    This PR (6282) - mean (1,107ms)  : 1084, 1130
     .   : milestone, 1107,
    master - mean (1,110ms)  : 1091, 1129
     .   : milestone, 1110,

```
```mermaid
gantt
    title Execution time (ms) FakeDbCommand (.NET Core 3.1) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (6282) - mean (108ms)  : 106, 110
     .   : milestone, 108,
    master - mean (109ms)  : 106, 111
     .   : milestone, 109,

    section CallTarget+Inlining+NGEN
    This PR (6282) - mean (764ms)  : 752, 777
     .   : milestone, 764,
    master - mean (765ms)  : 747, 783
     .   : milestone, 765,

```
```mermaid
gantt
    title Execution time (ms) FakeDbCommand (.NET 6) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (6282) - mean (92ms)  : 89, 94
     .   : milestone, 92,
    master - mean (92ms)  : 90, 94
     .   : milestone, 92,

    section CallTarget+Inlining+NGEN
    This PR (6282) - mean (718ms)  : 698, 737
     .   : milestone, 718,
    master - mean (726ms)  : 709, 742
     .   : milestone, 726,

```
```mermaid
gantt
    title Execution time (ms) HttpMessageHandler (.NET Framework 4.6.2) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (6282) - mean (191ms)  : 187, 195
     .   : milestone, 191,
    master - mean (191ms)  : 184, 199
     .   : milestone, 191,

    section CallTarget+Inlining+NGEN
    This PR (6282) - mean (1,226ms)  : 1200, 1252
     .   : milestone, 1226,
    master - mean (1,228ms)  : 1207, 1249
     .   : milestone, 1228,

```
```mermaid
gantt
    title Execution time (ms) HttpMessageHandler (.NET Core 3.1) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (6282) - mean (276ms)  : 270, 283
     .   : milestone, 276,
    master - mean (277ms)  : 273, 281
     .   : milestone, 277,

    section CallTarget+Inlining+NGEN
    This PR (6282) - mean (942ms)  : 925, 958
     .   : milestone, 942,
    master - mean (943ms)  : 925, 961
     .   : milestone, 943,

```
```mermaid
gantt
    title Execution time (ms) HttpMessageHandler (.NET 6) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (6282) - mean (266ms)  : 260, 272
     .   : milestone, 266,
    master - mean (266ms)  : 262, 270
     .   : milestone, 266,

    section CallTarget+Inlining+NGEN
    This PR (6282) - mean (925ms)  : 907, 942
     .   : milestone, 925,
    master - mean (931ms)  : 914, 947
     .   : milestone, 931,

```

**Comment by andrewlock** [comment]
## Throughput/Crank Report :zap:

Throughput results for AspNetCoreSimpleController comparing the following branches/commits:
- [This PR (6282)](https://github.com/DataDog/dd-trace-dotnet/tree/42843f51a5c9f9fcee7ec09554ca49e4871ff160)
- [master](https://github.com/DataDog/dd-trace-dotnet/tree/6aa8c666b2a48b6c5a85044ce60a31a23c80970e)
- [benchmarks/2.9.0](https://github.com/DataDog/dd-trace-dotnet/tree/89ae3f4ee7d85b824badda5873166707c9e560b4)

Cases where throughput results for the PR are worse than latest master (5% drop or greater), results are shown in **red**.

Note that these results are based on a _single_ point-in-time result for each branch. For full results, see [one](https://ddstaging.datadoghq.com/dashboard/fnz-afi-c2i/apm-net-crank) of the [many](https://ddstaging.datadoghq.com/dashboard/c5g-i7d-kcy/ciapp-apm-net-throughput-tests), [many](https://ddstaging.datadoghq.com/dashboard/uxh-m8j-qhi/ciapp-apm-net-throughput-tests-kevin) dashboards!

```mermaid
gantt
    title Throughput Linux x64 (Total requests) 
    dateFormat  X
    axisFormat %s
    section Baseline
    This PR (6282) (11.193M)   : 0, 11192639
    master (11.115M)   : 0, 11115106
    benchmarks/2.9.0 (11.033M)   : 0, 11032866

    section Automatic
    This PR (6282) (7.308M)   : 0, 7307719
    master (7.198M)   : 0, 7197853
    benchmarks/2.9.0 (7.786M)   : 0, 7785853

    section Trace stats
    master (7.426M)   : 0, 7426248

    section Manual
    master (11.134M)   : 0, 11134495

    section Manual + Automatic
    This PR (6282) (6.753M)   : 0, 6753293
    master (6.562M)   : 0, 6561650

    section DD_TRACE_ENABLED=0
    master (10.182M)   : 0, 10182063

```
```mermaid
gantt
    title Throughput Linux arm64 (Total requests) 
    dateFormat  X
    axisFormat %s
    section Baseline
    This PR (6282) (9.496M)   : 0, 9496253
    master (9.510M)   : 0, 9510131
    benchmarks/2.9.0 (9.495M)   : 0, 9494821

    section Automatic
    This PR (6282) (6.527M)   : 0, 6527317
    master (6.393M)   : 0, 6393402

    section Trace stats
    master (6.677M)   : 0, 6677096

    section Manual
    master (9.581M)   : 0, 9581463

    section Manual + Automatic
    This PR (6282) (5.999M)   : 0, 5998619
    master (6.035M)   : 0, 6034859

    section DD_TRACE_ENABLED=0
    master (8.852M)   : 0, 8852473

```
```mermaid
gantt
    title Throughput Windows x64 (Total requests) 
    dateFormat  X
    axisFormat %s
    section Baseline
    This PR (6282) (9.667M)   : 0, 9667167
    master (9.696M)   : 0, 9696335
    benchmarks/2.9.0 (10.020M)   : 0, 10019592

    section Automatic
    This PR (6282) (6.611M)   : 0, 6610614
    master (6.737M)   : 0, 6736896
    benchmarks/2.9.0 (7.255M)   : 0, 7255257

    section Trace stats
    master (7.067M)   : 0, 7067258

    section Manual
    master (10.012M)   : 0, 10011917

    section Manual + Automatic
    This PR (6282) (6.262M)   : 0, 6261904
    master (6.326M)   : 0, 6326467

    section DD_TRACE_ENABLED=0
    master (8.846M)   : 0, 8845694

```

File: tracer/src/Datadog.Trace/Telemetry/Collectors/DependencyTelemetryCollector.cs [hunk 1]
```diff
@@ -176,15 +176,19 @@ private static bool IsHexChar(char c)
...
+            return (assemblyName.Length == 8
+                 && IsBase32Char(assemblyName[0])
+                 && IsBase32Char(assemblyName[1])
+                 && IsBase32Char(assemblyName[2])
+                 && IsBase32Char(assemblyName[3])
+                 && IsBase32Char(assemblyName[4])
+                 && IsBase32Char(assemblyName[5])
+                 && IsBase32Char(assemblyName[6])
+                 && IsBase32Char(assemblyName[7]))
+                || (assemblyName.Length == 32 && IsHexString(assemblyName, 0));
```
[the remark this answers is not in the document]
Comment (lucaspimentel) [reply tracer/src/Datadog.Trace/Telemetry/Collectors/DependencyTelemetryCollector.cs:191]: Or `>= 32` if you think `16` is not enough.

**Review by lucaspimentel — APPROVED** [review]

File: tracer/src/Datadog.Trace/Telemetry/Collectors/DependencyTelemetryCollector.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (andrewlock) [reply tracer/src/Datadog.Trace/Telemetry/Collectors/DependencyTelemetryCollector.cs:191]: Yeah, I toyed with that. I think 16+ is almost certainly going to be ok, but will use 32 for now, just for safety. In reality I wonder if we even need this check at all for Version = `0.0.0.0` 🤷‍♂️ Whatever we do, there's going to be a certain amount of whackamole. But I guess we're unlikely to have `ABadCafeFedDeafBeesBeef.dll` 😄 

```suggestion
                || (assemblyName.Length >= 32 && IsHexString(assemblyName, 0));
```

**Review by kevingosse — APPROVED** [review]

**Pull request merged**